zerver/tests/test_events.py

```diff
@@ -1586,6 +1586,43 @@ def test_change_avatar_fields(self) -> None:
         error = schema_checker('events[0]', events[0])
         self.assert_on_error(error)
 
+        # Test for exclude_old_users_avatars = True
+        users = [
+            self.example_user('iago'),
+            self.example_user('cordelia'),
+            self.example_user('ZOE'),
+            self.example_user('othello'),
+            self.example_user('prospero'),
+            self.example_user('aaron'),
+            self.example_user('polonius'),
+            self.example_user('desdemona'),
+        ]
+        # Set long_term_idle = True for some users. This list
+        # don't include the logged-in user i.e. Hamlet as the
+        # logged-in user can't be long term idle.
+        for user in users:
+            user.long_term_idle = True
+            user.save()
+        schema_checker = self.check_events_dict([
+            ('type', equals('realm_user')),
+            ('op', equals('update')),
+            ('person', check_dict_only([
+                ('avatar_url', check_string),
+                ('avatar_url_medium', check_string),
+                ('avatar_version', check_int),
+                ('avatar_source', check_string),
+                ('user_id', check_int),
+            ])),
+        ])
+        # In order to test exclude_old_users_avatars effectively,
+        # we have to set client_gravatar = False
+        events = self.do_test(
+            lambda: do_change_avatar_fields(self.user_profile, UserProfile.AVATAR_FROM_USER),
+            exclude_old_users_avatars=True, client_gravatar=False
+        )
+        error = schema_checker('events[0]', events[0])
+        self.assert_on_error(error)
+
```

Comment:
I think that you should remove this chunk entierly since it's not really testing anything and since it seems like a partial and incomplete duplicate of the test you added below (`test_exclude_old_users_avatars`). It feels like you meant to remove this once you added `test_exclude_old_users_avatars` but then forgot?

Reply:
Oh wait, was this the test for `apply_events` that you were talking about? If so you did this wrong.

The lambda function that you pass to `do_test` should be the event you want to check for consistency around - it's the action that you want to verify doesn't make the normal state and hybrid state different. Here you've made that action `do_change_avatar_fields` which is not related to changing `long_term_idle`. This is what makes it so that you're not testing anything new here (you're actually testing the wrong thing). The function you pass should be toggling the `long_term_idle` of the user (which you tried to did preemptively).

Also, if you want to make `long_term_idle=True` for everyone then shouldn't `avatar_url` be checking None instead of string? In which case `('avatar_url', check_string)` is wrong.

On a somewhat related note, Take a quick look at [this](https://github.com/zulip/zulip/pull/12554#issuecomment-500949159) comment about some performance optimizations that `do_test` uses. It's something that you should keep in mind for any future tests you add involving `do_test` I remember that it drove me crazy last summer (that and Python's default mutable arguments feature).

Reply:
So I have looked into it further, and I think that there is no event that triggers the soft deactivation of users. Can we still add tests for this?

Reply:
Correct, soft deactivation is, by design, a concept that is invisible to clients. 

We're breaking that rule a little bit with this change, but the API structure is that the server is allowed to choose freely which users to send avatar URLs for, and which to not do so, entirely at its discretion if the client declares it has support for this.  (And we may later change the algorithm; `long_term_idle` just seems like a good place to start)

Reply:
@hashirsarwar I think that you can write a test method to manually soft deactivated certain user by just setting `lomg_term_idle` to `True`.

Reply:
yeah, agreed